### PR TITLE
See #7777. Use vhost_max to stop vhost creation in rabbitmq

### PR DESCRIPTION
**Pull request by SimonUnge**
## Proposed Changes

See #7777 

Limit added for number of vhost that can be created.
I *think* I have added proper handling so the UI gets a nice popup, and the ctl gets a nice error format.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

_Put an `x` in the boxes that apply.
You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask on the mailing list.
We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with my changes
- [x] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [x] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

### CODE REVIEW

File: deps/rabbitmq_management/src/rabbit_mgmt_wm_vhost.erl [hunk 1]
```diff
@@ -60,25 +60,28 @@ accept_content(ReqData0, Context = #context{user = #user{username = Username}})
...
+              Description = maps:get(description, BodyMap, <<"">>),
+              Tags = maps:get(tags, BodyMap, <<"">>),
+              %% defaultqueuetype was an unfortunate name picked originally for 3.11.0,
+              %% so fall back to it. See rabbitmq/rabbitmq-server#7734.
+              FallbackQT = maps:get(defaultqueuetype, BodyMap, undefined),
+              DefaultQT = maps:get(default_queue_type, BodyMap, FallbackQT),
+              case put_vhost(Name, Description, Tags, DefaultQT, Trace, Username) of
+                  ok ->
+                      {true, ReqData, Context};
+                  {'EXIT', #amqp_error{name = precondition_failed,
```
Comment (SimonUnge) [inline deps/rabbitmq_management/src/rabbit_mgmt_wm_vhost.erl:73]: the amqp_error comes from `rabbit_misc:protocol_error`. Maybe there is a better structure to use?

File: deps/rabbit/src/rabbit_vhost.erl [hunk 2]
```diff
@@ -331,6 +333,23 @@ put_vhost(Name, Description, Tags0, DefaultQueueType, Trace, Username) ->
...
+    is_over_vhost_limit(Name, Limit).
+
+is_over_vhost_limit(_Name, infinity) ->
+    ok;
+is_over_vhost_limit(Name, Limit) when is_integer(Limit) ->
+    case length(rabbit_db_vhost:list()) >= Limit of
+        false ->
+            ok;
+        true ->
+            rabbit_misc:protocol_error(precondition_failed,
```
Comment (michaelklishin) [inline deps/rabbit/src/rabbit_vhost.erl:347]: We cannot use `rabbit_misc:protocol_error/3` here: it is meant to be used in the context of a connection or channel, so, to communicate errors to clients.

**Review by michaelklishin — CHANGES_REQUESTED** [review]

File: deps/rabbit/src/rabbit_vhost.erl (hunk 2 above)
Comment (SimonUnge) [reply deps/rabbit/src/rabbit_vhost.erl:347]: Yeah, makes sense. On it.

**Comment by SimonUnge** [comment]
@michaelklishin updated to use a 'custom' vhost error.

**Pull request merged**